
# Tutorial: Parameters vs. Hyperparameters: The Crucial Difference

Selamat datang di Chapter 3! Sebelum kita terjun ke teknik *tuning* yang canggih, kita harus memahami dengan jelas **apa** yang sebenarnya akan kita *tuning*. Dalam *machine learning*, ada dua jenis "pengaturan" model yang seringkali membingungkan bagi pemula: **Parameters** dan **Hyperparameters**.

Memahami perbedaan keduanya adalah kunci untuk menjadi praktisi *machine learning* yang efektif.

Di notebook ini, kita akan membedah perbedaan fundamental ini dengan analogi dan contoh kode yang jelas.




---
### 1. Tujuan Pembelajaran

Di akhir notebook ini, Anda akan dapat:

* Mendefinisikan apa itu **Parameter** model.
* Mendefinisikan apa itu **Hyperparameter** model.
* Memberikan contoh konkret untuk keduanya menggunakan model Scikit-learn.
* Menjelaskan mengapa kita hanya bisa "menyetel" (*tune*) hyperparameter, bukan parameter.




---
### 2. Analogi: Koki dan Resepnya

Bayangkan sebuah model *machine learning* sebagai seorang koki yang cerdas.

* **Hyperparameters** adalah **keputusan tingkat tinggi** yang dibuat oleh *Anda* (sang kepala koki) **sebelum** memasak dimulai. Contoh: "Gunakan oven dengan suhu 200°C", "Panggang selama 25 menit", "Gunakan resep A, bukan resep B". Ini adalah **aturan main** yang Anda berikan kepada koki.

* **Parameters** adalah **wawasan atau "rasa"** yang **dipelajari oleh koki itu sendiri** selama proses memasak (pelatihan). Contoh: Berdasarkan pengalaman mencicipi, koki menemukan bahwa rasio garam dan merica yang paling pas adalah 2:1. Anda tidak memberitahunya rasio ini; ia **menemukannya sendiri** dari "data" (bahan-bahan).

**Intinya:** Kita mengatur *hyperparameter*, dan model belajar *parameter*.




---
### 3. Setup: Data dan Model Sederhana

Kita akan menggunakan dataset sederhana dan dua model untuk demonstrasi: `LinearRegression` (yang tidak punya hyperparameter penting) dan `Ridge` (yang punya satu hyperparameter penting).



In [4]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Ridge

# membuat dataset
np.random.seed(42)
X = np.random.rand(100,3) * 10
y = 2 * X[:,0] + 3 * X[:,1] - 5 * X[:,2] + np.random.randn(100)*2

X_train,X_test,y_train,y_test = train_test_split(X,y,random_state=42)


---
### 4. Parameters: Apa yang Model Pelajari

Parameter adalah nilai-nilai internal dari model yang ditemukan selama proses pelatihan (`.fit()`). Kita tidak bisa mengaturnya secara langsung.

Untuk Regresi Linear, parameternya adalah **koefisien (`coef_`)** dan **intersep (`intercept_`)**.



In [5]:
lr = LinearRegression()
lr.fit(X_train,y_train)

# periksa parameter
print("Parameter yang dipelajari oleh linear regression")
print(f" - intercept (beta_0) = {lr.intercept_:.4f}")
print(f" - koefisien (beta_1,beta_2,beta_3) = {np.round(lr.coef_,4)}")

Parameter yang dipelajari oleh linear regression
 - intercept (beta_0) = -0.1446
 - koefisien (beta_1,beta_2,beta_3) = [ 2.0347  2.9877 -4.8751]



**Observasi:**

Nilai-nilai di atas (`-0.1446` dan `[ 2.0347  2.9877 -4.8751]`) tidak kita tentukan. Model **menemukan** nilai-nilai ini dari data `X_train` dan `y_train` sebagai "garis paling pas" terbaik. Inilah yang disebut **parameter**.




---
### 5. Hyperparameters: Apa yang Kita Atur

Hyperparameter adalah "tombol" yang kita putar **sebelum** melatih model untuk mengontrol perilakunya.

Mari kita gunakan `Ridge Regression`. Model ini sangat mirip dengan Regresi Linear, tetapi ia memiliki satu hyperparameter penting: `alpha`. `alpha` mengontrol seberapa besar kita "menghukum" model karena memiliki koefisien yang besar (ini adalah teknik regularisasi yang akan kita pelajari nanti).

Mari kita latih dua model `Ridge` dengan nilai `alpha` yang berbeda.



In [6]:
# model 1: dengan hyperparameter alpha yang kecil
ridge_low_alpha = Ridge(alpha=0.1)
ridge_low_alpha.fit(X_train,y_train)

# model 2: dengan hyperparameter alpha yang besar
ridge_high_alpha = Ridge(alpha=100)
ridge_high_alpha.fit(X_train,y_train)

print("Parameter yang dipelajari oleh ridge alpha 0.1")
print(f" - koefisien yang dipelajari = {np.round(ridge_low_alpha.coef_,4)}")
print("Parameter yang dipelajari oleh ridge alpha 100")
print(f" - koefisien yang dipelajari = {np.round(ridge_high_alpha.coef_,4)}")

Parameter yang dipelajari oleh ridge alpha 0.1
 - koefisien yang dipelajari = [ 2.0344  2.9874 -4.8743]
Parameter yang dipelajari oleh ridge alpha 100
 - koefisien yang dipelajari = [ 1.789   2.6551 -4.2184]



**Observasi Kunci:**

Lihat bagaimana **mengubah hyperparameter (`alpha`)** secara langsung **mengubah parameter (`coef_`)** yang dipelajari oleh model!

* Dengan `alpha` yang rendah, koefisiennya hampir sama dengan Regresi Linear biasa.
* Dengan `alpha` yang tinggi, model "dihukum" dan dipaksa untuk mengecilkan koefisiennya.

Inilah inti dari *hyperparameter tuning*: kita mencoba berbagai nilai untuk hyperparameter (`alpha`) untuk menemukan set parameter (`coef_`) yang menghasilkan model dengan performa terbaik.




---
### 6. Ringkasan: Tabel Perbandingan

| Fitur | Parameters | Hyperparameters |
| :--- | :--- | :--- |
| **Siapa yang Mengatur?** | **Model** | **Pengguna (Anda)** |
| **Kapan Diatur?** | Selama proses `.fit()` (pelatihan) | **Sebelum** proses `.fit()` |
| **Sumber Nilai?** | Berasal dari data | Berasal dari coba-coba, pengalaman, atau pencarian otomatis |
| **Contoh** | `coef_`, `intercept_` | `alpha`, `n_estimators`, `max_depth`, `C` |
| **Tujuan** | Untuk membuat prediksi | Untuk mengontrol proses pembelajaran model |




### 7. Kesimpulan Tutorial

* **Parameters** dipelajari dari data.
* **Hyperparameters** diatur oleh kita.
* Tujuan kita dalam *hyperparameter tuning* adalah **menemukan nilai hyperparameter terbaik** yang akan memandu model untuk **mempelajari parameter terbaik**.

Di subchapter selanjutnya, kita akan belajar cara mencari nilai hyperparameter terbaik ini secara otomatis menggunakan `GridSearchCV`.
